### PR TITLE
Always tick the QUIC_ENGINE (for the quic-server branch)

### PR DESCRIPTION
**Pull request by mattcaswell**
This PR will eventually target the quic-server branch - but is currently built on top of #25432 and is pending the merge of that PR. Only the last 2 commits are relevant.

Just because one connection has not started yet, it does not mean that we should not tick the QUIC_ENGINE. There may be other connections that do need ticking.

We also add a test for the issue described in #25054, which was originally fixed by #25069 to ensure we haven't broken anything by the changes in the other commit in this PR.

### CODE REVIEW

**Comment by mattcaswell** [comment]
Setting this as draft until the rebase PR has been merged. Just look at the last two commits.

@vdukhovni, @t8m, @Sashan

File: ssl/quic/quic_channel.c
Comment (Sashan) [inline ssl/quic/quic_channel.c:1877]: may be we should have a function (`ossl_quic_channel_is_idle()`) for this, just to make code consistent with style set by Hugo. This is cosmetic nit, I don't insist on.

**Comment by vdukhovni** [comment]
Looks good so far.  The remaining todo is about the synthetic clock moving from QC to the engine, simplifying further the upper-layer code.

**Review by Sashan — CHANGES_REQUESTED** [review]
looks good I have just one nit suggestion to consider.

File: ssl/quic/quic_impl.c
Comment (mattcaswell) [reply ssl/quic/quic_impl.c:1304]: Yes - because they are no longer needed. The quic server branch handles this in a different way (with a check whether the channel has started in the `ossl_quic_channel_subtick` function, which does not exist in the master branch).

File: ssl/quic/quic_channel.c
Comment (mattcaswell) [reply ssl/quic/quic_channel.c:1877]: I don't particularly feel the need for this.

**Comment by nhorman** [comment]
waiving the style here, since its not relevant to the two pertinent commits, but we should maybe take a pass through the code after the rebase to address the style issues prior to merge

**Review by Sashan — APPROVED** [review]
looks good to me.

**Comment by mattcaswell** [comment]
> The remaining todo is about the synthetic clock moving from QC to the engine, simplifying further the upper-layer code.

There does seem to be some support for this already:

https://github.com/openssl/openssl/blob/a6fe4522996b4dd751b4ee0f12982d84430e3e1b/ssl/quic/quic_engine_local.h#L43-L45

**Review by nhorman — APPROVED** [review]

**Comment by vdukhovni** [comment]
> > The remaining todo is about the synthetic clock moving from QC to the engine, simplifying further the upper-layer code.
> 
> There does seem to be some support for this already:
> 
> https://github.com/openssl/openssl/blob/a6fe4522996b4dd751b4ee0f12982d84430e3e1b/ssl/quic/quic_engine_local.h#L43-L45

So is there still any need in your view to have a separate clock override at the connection layer, or is that now obsolete?
I'd be tempted to migrate all use of the connection clock to the engine clock, simplifying the timeout code to be connection-type independent.

**Comment by mattcaswell** [comment]
> So is there still any need in your view to have a separate clock override at the connection layer, or is that now obsolete?

It's still used, and it would be some work to remove it. But it does make sense to me for the clock to be at the engine level. But I think that's a different PR to this one.

**Comment by vdukhovni** [comment]
> > So is there still any need in your view to have a separate clock override at the connection layer, or is that now obsolete?
> 
> It's still used, and it would be some work to remove it. But it does make sense to me for the clock to be at the engine level. But I think that's a different PR to this one.

Is it possible to "transparently" retarget the override from the QC to the engine layer?

**Comment by mattcaswell** [comment]
> Is it possible to "transparently" retarget the override from the QC to the engine layer?

Actually looking further at this it seems to me that virtually all the required plumbing for this is already present. The QUIC_CHANNEL `get_time` function just calls `ossl_quic_port_get_time(ch->port)` and that function just calls `ossl_quic_engine_get_time(port->engine)`

The only odd part is that you have to set the override via `ossl_quic_conn_set_override_now_cb` which calls `expect_quic_conn_only` and stores the function pointer in the `QUIC_CONNECTION` object. But it still wraps that function pointer and sets it at the QUIC_ENGINE level.

So it looks to me like everything we need for now is already there. At some point we probably want to change (and rename) `ossl_quic_conn_set_override_now_cb`  so that it instead calls `expect_quic_any` and stores the function pointer at the QUIC_ENGINE level instead. But we don't need to do that until such time as we want to set an override on a non-connection based SSL object (listener or domain).

**Comment by vdukhovni** [comment]
> > Is it possible to "transparently" retarget the override from the QC to the engine layer?
> 
> Actually looking further at this it seems to me that virtually all the required plumbing for this is already present. The QUIC_CHANNEL `get_time` function just calls `ossl_quic_port_get_time(ch->port)` and that function just calls `ossl_quic_engine_get_time(port->engine)`
> 
> The only odd part is that you have to set the override via `ossl_quic_conn_set_override_now_cb` which calls `expect_quic_conn_only` and stores the function pointer in the `QUIC_CONNECTION` object. But it still wraps that function pointer and sets the wrapped function at the QUIC_ENGINE level.
> 
> So it looks to me like everything we need for now is already there. At some point we probably want to change (and rename) `ossl_quic_conn_set_override_now_cb` so that it instead calls `expect_quic_any` and stores the function pointer at the QUIC_ENGINE level instead. But we don't need to do that until such time as we want to set an override on a non-connection based SSL object (listener or domain).

Not in this PR, but I'd probably like to see it sooner, so that all the code that handles time is uniform and simpler. So just asking while this is under the microscope, but sure one thing at a time. Perhaps the clock cleanup will come soon enough.

**Comment by t8m** [comment]
@mattcaswell Can you please rebase and retargett this PR appropriately?

**Comment by mattcaswell** [comment]
> Not in this PR, but I'd probably like to see it sooner, so that all the code that handles time is uniform and simpler. So just asking while this is under the microscope, but sure one thing at a time. Perhaps the clock cleanup will come soon enough.

#25457

**Comment by mattcaswell** [comment]
Rebased, retargeted and taken out of draft. I also cleared the "style: waived" label to see if there are any style issues.

Please review.

**Review by Sashan — APPROVED** [review]
still looks good after rebase.

**Comment by mattcaswell** [comment]
Still seeking a second approval for this: @openssl/committers

**Comment by mattcaswell** [comment]
CI failure doesn't look relevant - although not one I've seen before?

**Review by vdukhovni — APPROVED** [review]

**Comment by vdukhovni** [comment]
> CI failure doesn't look relevant - although not one I've seen before?

The command in question is: https://github.com/open-quantum-safe/oqs-provider/blob/4db09a9dc540543ff0e22b2713757a7e90e1f0c6/scripts/oqsprovider-externalinterop.sh#L41-L42

**Comment by vdukhovni** [comment]
> > CI failure doesn't look relevant - although not one I've seen before?
> 
> The command in question is: https://github.com/open-quantum-safe/oqs-provider/blob/4db09a9dc540543ff0e22b2713757a7e90e1f0c6/scripts/oqsprovider-externalinterop.sh#L41-L42

The hostname to connect to and the SNI name are surprisingly in conflict. Is that correct?

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**
